torch/_inductor/codegen/triton_foreach.py

```diff
@@ -1,81 +1,140 @@
-        cur_partition = []
-        for node in nodes:
+    def _update_partition(partition_state, node_rw_count, node_info):
+        if partition_state.cur_count + node_rw_count > ForeachKernel.MAX_NUM_ARGS:
+            partition_state.partitions.append(partition_state.cur_partition)
+            partition_state.cur_partition = [node_info]
+            partition_state.cur_count = node_rw_count
+        else:
+            partition_state.cur_count += node_rw_count
+            partition_state.cur_partition.append(node_info)
+
+    @staticmethod
+    def horizontal_partition(subkernel_nodes, triton_scheduling):
+        """Generates a list of lists of node info tuples which consist of (fused_nodes, tiling, numel, rnumel)
+        for each subkernel node where each sublist is guaranteed to not exceed CUDA limits for number of args
+        (read/writes) and to have the same 2D or 1D blocking strategy."""
+        assert len(subkernel_nodes) >= 1
+
+        partition_state_1d = PartitionState([], [], 0)
+        yelem_to_partition_state_2d = defaultdict(lambda: PartitionState([], [], 0))
+
+        for node in subkernel_nodes:
+            fused_nodes = node.get_nodes()
+            _, (numel, rnumel) = max(
+                fused_nodes, key=lambda x: int(x.is_reduction())
+            ).group
+            tiled_groups = triton_scheduling.select_tiling(fused_nodes, numel, rnumel)
+            node_info = fused_nodes, tiled_groups, numel, rnumel
+
             read_writes = node.read_writes
             read_write_count = len(read_writes.reads) + len(read_writes.writes)
-            if cur_count + read_write_count > ForeachKernel.MAX_NUM_ARGS:
-                partitions.append(cur_partition)
-                cur_partition = [node]
-                cur_count = read_write_count
+
+            if tiled_groups[1] == 1:
+                ForeachKernel._update_partition(
+                    partition_state_1d, read_write_count, node_info
+                )
             else:
-                cur_count += read_write_count
-                cur_partition.append(node)
+                y_elem = tiled_groups[0]
+                partition_state_2d = yelem_to_partition_state_2d[y_elem]
+                ForeachKernel._update_partition(
+                    partition_state_2d, read_write_count, node_info
+                )
 
-        if cur_partition:
-            partitions.append(cur_partition)
+        partition_state_1d.finalize()
+        all_partitions = partition_state_1d.partitions
+        for partition_state_2d in yelem_to_partition_state_2d.values():
+            partition_state_2d.finalize()
+            all_partitions.extend(partition_state_2d.partitions)
 
-        return partitions
+        return all_partitions
 
     def __init__(self):
         super().__init__()
-        self.block_size = 1024  # Try tuning this value
+        self.blocking_2d = False
+        self.block_size_1d = 1024  # Try tuning this value
+        self.block_size_2d = 32
```

Comment:
could we use the same autotuning we do for other kernels for foreach?  I worry 32 won't be the right block size.

Reply:
Yeah I think we can, do I just add it to the meta params?

Reply:
you might need to change the decorator